### PR TITLE
Comments: Try Threaded View

### PR DESCRIPTION
**Pull request by Copons**
WIP

### Testing instructions

- Find a comment belonging to a post with plenty of comments and replies.
- Click that comment's post title and then change the filter status to All.

### CODE REVIEW

File: client/my-sites/comments/comment-threaded-list/threaded-comment.jsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+				<Comment { ...{ commentId } } isPostView refreshCommentData />
+
+				{ !! replyTree.length &&
+					! isExpanded && (
+						<Notice
+							className="comment-threaded-list__more-replies"
+							icon="reply"
+							showDismiss={ false }
+						>
+							<a onClick={ this.showReplies }>{ translate( 'Show replies' ) }</a>
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-threaded-list/threaded-comment.jsx]: Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Maximum replies to show:' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-threaded-list/threaded-comment.jsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+		return (
+			<div className={ `comment-threaded-list__comment depth-${ depth }` }>
+				<Comment { ...{ commentId } } isPostView refreshCommentData />
+
+				{ !! replyTree.length && (
+					<ConnectedThreadedComment
+						commentId={ replyTree[ 0 ].commentId }
+						commentsTree={ commentsTree }
+						depth={ depth + 1 }
```
Comment (rodrigoi) [inline client/my-sites/comments/comment-threaded-list/threaded-comment.jsx]: good call. Traversing the comments tree to calculate depth was not a good idea.

**Comment by gwwar** [comment]
Just wanted to note that the Clicking on the post link goes to /approved. I wonder if we should change that to /all?

**Comment by gwwar** [comment]
cc @drw158 @megsfulton feel free to take this for an early spin. 

I verified that we only display the parent comment + 1 level deep in Reader, so behavior matches here.

Some Design ❓ 
1. Does the "In Reply To" become redundant in this view?
2. If we have long comments, will folks be able to tell that it's a reply?
3. Does it matter if the load more prompt follows the width of the comment above? I saw instances where it did/did not match.

screen shot 2017-11-21 at 2 22 04 pm

File: client/my-sites/comments/comment-threaded-list/index.jsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+export class CommentThreadedList extends Component {
+	static propTypes = {
+		postId: PropTypes.number,
+		siteFragment: PropTypes.string,
+		siteId: PropTypes.number,
+	};
+
+	state = {
+		commentsPerPage: 10,
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/index.jsx]: Do we still have pages in this view? Or is this max comments to load?

File: client/my-sites/comments/comment-threaded-list/index.jsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		pageLoaded: 1,
+	};
+
+	getRootComments = () => filter( this.props.commentsTree, { commentParentId: 0 } );
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/index.jsx]: Feels like this can be moved as a selector or at least into mapStateToProps. Eg we only need to calculate this when commentsTree is updated.

File: client/my-sites/comments/comment-threaded-list/index.jsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+	loadMoreComments = () =>
+		this.setState( ( { pageLoaded } ) => ( { pageLoaded: pageLoaded + 1 } ) );
+
+	renderLoadMore = () => {
+		const { translate } = this.props;
+		const { commentsPerPage, pageLoaded } = this.state;
+		const rootTotal = this.getRootComments().length;
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/index.jsx]: Let's pass this as a prop in mapStateToProps, we end up calling getRootComments() twice per render.

File: client/my-sites/comments/comment-threaded-list/index.jsx (hunk 4 above)
Comment (gwwar) [reply client/my-sites/comments/comment-threaded-list/index.jsx]: An alternative would be to create a new endpoint or update the comments-tree endpoint to return this for us.

File: client/my-sites/comments/comment-threaded-list/index.jsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+				{ this.renderLoadMore() }
+			</div>
+		);
+	}
+}
+
+const mapStateToProps = ( state, { postId, siteId } ) => {
+	const siteCommentsTree = getSiteCommentsTree( state, siteId, 'all' );
+	const commentsTree = sortBy( filter( siteCommentsTree, { postId } ), 'commentId' );
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/index.jsx]: If we end up doing this work on the client, we can memoize this.

File: client/my-sites/comments/comment-threaded-list/threaded-comment.jsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+		isExpanded: false,
+	};
+
+	showReplies = () => this.setState( { isExpanded: true } );
+
+	render() {
+		const { commentId, commentsTree, depth, isLoading, siteId, translate } = this.props;
+		const { isExpanded } = this.state;
+
+		const replyTree = filter( commentsTree, { commentParentId: commentId } );
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/threaded-comment.jsx]: Same thing, we should make this a selector, or at least move to mapStateToProps

File: client/my-sites/comments/comment-threaded-list/threaded-comment.jsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+		return (
+			<div className={ `comment-threaded-list__comment depth-${ depth }` }>
+				<Comment { ...{ commentId } } isPostView refreshCommentData />
+
+				{ !! replyTree.length && (
+					<ConnectedThreadedComment
+						commentId={ replyTree[ 0 ].commentId }
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/threaded-comment.jsx]: Maybe pass this as a prop:
```
const threadedCommentId = replyTree[ 0 ].commentId;
```

File: client/my-sites/comments/comment-threaded-list/threaded-comment.jsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+						commentsTree={ commentsTree }
```
Comment (gwwar) [inline client/my-sites/comments/comment-threaded-list/threaded-comment.jsx]: We probably want to specify that this isn't the full comment's tree

**Comment by gwwar** [comment]
Thanks for putting this together @Copons! It'll be really handy for us to be able to interact with this view. 

@rodrigoi you mentioned that you were seeing perf issues on your test site. What does your test data look like, so we can help reproduce?

**Comment by rodrigoi** [comment]
For development, I use a blog post with exactly 1645 comments. Some are replies, most are just plain comments.
I tried this branch with that post, and it took it a while to load. It eventually did, but performance degradation was noticeable.
The same happened with my experiment on https://github.com/Automattic/wp-calypso/pull/20077, and it benefited a lot from some memoization.

BTW, that PR does the same with a smaller diff 😛

**Comment by Copons** [comment]
> BTW, that PR does the same with a smaller diff 😛 

@rodrigoi well not really, I've got the load more parts! 😄

**Comment by Copons** [comment]
Thanks for the review folks!

> 1. Does the "In Reply To" become redundant in this view?

@gwwar Not really. We've got 2 depth level tops, so "replies of replies" aren't nested.

> 2. If we have long comments, will folks be able to tell that it's a reply?

Probably not? I willingly avoided to take many design decisions because imho right now is more important to get the logic right.
We could add a "load more" on the comments content, though.

> 3. Does it matter if the load more prompt follows the width of the comment above? I saw instances where it did/did not match.
>
> screen shot 2017-11-21 at 2 22 04 pm


That's a sort of glitch caused by my "load 1 reply" approach.
If the parent has more replies, but that reply doesn't, we see that weird missing padding.

In other words, in that screen: the "Load more" doesn't belong to the comment right above, but to its parent.
So its (lack of) padding is correct, but it definitely looks wrong.

**Comment by gwwar** [comment]
Let's see, maybe let's close this one out and have folks pair together (/async) on https://github.com/Automattic/wp-calypso/pull/20077. Just remember not to force rebase 😄  without a heads up.

I think visually this prototype is pretty close to the mocks ✨, but @rodrigoi has a bit more of the functionality displayed 🛠. With your powers combined, I'm sure it'll turn out great.

**Comment by Copons** [comment]
Closed in favour of #20077
